src/toil/job.py

```diff
@@ -1722,3 +1721,185 @@ def convertPromises(kwargs):
+             These defaults are removed if docker_parmaters is passed, so be sure to pass them if they are desired.
+    :param file outfile: Pipe output of Docker call to file handle
+    :param bool check_output: When True, this function returns docker's output.
+    :param int defer: What action should be taken on the container upon job completion?
+           docker_call.FORGO (0) will leave the container untouched.
+           docker_call.STOP (1) will attempt to stop the container with `docker stop` (useful for
+           debugging).
+           docker_call.RM (2) will stop the container and then forcefully remove it from the system
+           using `docker rm -f`.
+           The default value is None and that shadows docker_call.FORGO, unless --rm is being passed in.
+    """
+    docker_call.FORGO = 0
+    docker_call.STOP = 1
+    docker_call.RM = 2
+
+    if parameters is None:
+        parameters = []
+
+    # Setup the outgoing subprocess call for docker
+    base_docker_call = ['docker', 'run']
+    if docker_parameters:
+        base_docker_call += docker_parameters
+    else:
+        base_docker_call += ['--rm', '--log-driver', 'none', '-v',
+                             os.path.abspath(work_dir) + ':/data']
+
+    # Ensure the user has passed a valid value for defer
+    require(defer in (None, docker_call.FORGO, docker_call.STOP, docker_call.RM),
+            'Please provide a valid value for defer.')
+
+    # Get container name which is needed for _docker_kill
+    try:
+        if any('--name=' in x for x in base_docker_call):
+            container_name = [x.split('=')[1] for x in base_docker_call if '--name=' in x][0]
+        else:
+            container_name = base_docker_call[base_docker_call.index('--name') + 1]
+    except ValueError:
+        container_name = _get_container_name(job)
+        base_docker_call.extend(['--name', container_name])
+    except IndexError:
+        raise RuntimeError("Couldn't parse Docker's `--name=` option, check parameters: " + str(base_docker_call))
+
+    # Defer the container on-exit action
+    if '--rm' in base_docker_call and defer is None:
+        defer = docker_call.RM
+    job.defer(_docker_kill, container_name, action=defer)
+    # Defer the permission fixing function.  We call this explicitly later on in this function, but
+    # we defer it as well to handle unexpected job failure.
+    job.defer(_fix_permissions, base_docker_call, tool, work_dir)
+
+    # Make subprocess call
+    call = base_docker_call + [tool] + parameters
+    job.fileStore.logToMaster("Calling docker with {}.".format(' '.join(call)))
+
+    if outfile:
+        subprocess.check_call(call, stdout=outfile)
+    else:
+        if check_output:
+            return subprocess.check_output(call)
+        else:
+            subprocess.check_call(call)
+
+    # Fix root ownership of output files
+    _fix_permissions(base_docker_call, tool, work_dir)
```

Comment:
Ya, when the deferred code was added a comment was left ,"We call this explicitly later on in this function, but we defer it as well to handle unexpected job failure."  so I left thinking it was intentional, but if the deferred function is always called then it's unnecessary to also include it at the end of the function.